File: edit distance dynamic programming.jl

# coding: utf-8

# In[1]:


#its also called levenshtein distance
#it can be done via recursion too
#the recursion version is much more elegant yet less efficient
# https://github.com/je-suis-tm/recursion/blob/master/edit%20distance%20recursion.jl

#edit distance is to minimize the steps transforming one string to another
#the way to solve this problem is very similar is to knapsack
#assume we have two strings text1 and text2
#we build a matrix with the size of (length(text1)+1)*(length(text2)+1)

#there are three different ways to transform a string
#insert,delete and replace
#we can use any of them or combined
#lets take a look at the best case first
#assume string text1 is string text2
#we dont need to do anything
#so 0 steps would be the answer
#for the worst case
#when string text1 has nothing in common with string text2
#we would have to replace the whole string text1
#the steps become the maximum step which is max(length(text1),length(text2))
#for general case,the number of steps would fall between the worst and the best case
#assume we are at i th letter of string text1 and j th letter string text2
#if we wanna get the optimal steps of transforming string text1 to string text2
#we have to make sure at each letter transformation
#text1[1:i] and text2[1:j] have reached their optimal status
#otherwise,we could always find another combination of insert,delete and replace
#to get a "real" optimal text1[1:i] and text2[1:j]
#it would make our string transformation not so optimal any more
#it is the same logic as the optimization of knapsack problem
#after we set our logic straight
#we would take a look at three different approaches
#lets take a look at insertion
#basically we need to insert j th letter from string text2 into string text1 at i th position
#the cumulated steps we have taken should be matrix[i][j-1]+1
#matrix[i][j-1] is the steps for text1[1:i] to text2[1:j]
#for delete,it is vice versa
#the cumulated steps we have taken should be matrix[i-1][j]+1
#for replacement,it is a lil bit tricky
#there are two scenarios
#if text1[i-1]==text2[j-1]
#it should be matrix[i-1][j-1]
#we dont need any replacement at all
#else,it should be matrix[i-1][j-1]+1
#we replace i th letter of string text1 with j th letter of string text2
#after we managed to understand three different approaches
#we want to take the minimum number of steps among these three approaches
#throughout the iteration of different positions of both strings
#in the end,we would get the optimal steps to transform one string to another,YAY


# In[2]:


function edit_distance(text1,text2)
    
    len1=length(text1)+1
    len2=length(text2)+1
    
    #this part is to create a matrix of (length(text1)+1)*(length(text2)+1)
    matrix=[[0 for _ in 1:len2] for _ in 1:len1]
    
    for i in 1:len1
        
        matrix[i][1]=i-1
    
    end
    
    for i in 1:len2
        
        matrix[1][i]=i-1
        
    end
    
    #we take iterations on both string text1 and text2
    #next,we check if text1[i-1]==text2[j-1]
    #if yes,no replacement needed
    #if no,replacement needed
    #we take a minimum function to see which combination would give the minimum steps
    #eventually we got what we are after
    for i in 2:len1
        
        for j in 2:len2
            
            if text1[i-1]==text2[j-1]
                
                matrix[i][j]=min(matrix[i-1][j]+1,
                matrix[i][j-1]+1,
                matrix[i-1][j-1])
                
            else
                
                matrix[i][j]=min(matrix[i-1][j]+1,
                matrix[i][j-1]+1,
                matrix[i-1][j-1]+1)
                
            end
            
        end
        
    end
    
    return matrix[len1][len2]
    
end


# In[3]:


println(edit_distance("baiseé","bas"))

File: edit distance dynamic programming.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 19 15:22:38 2018

@author: Administrator
"""

#its also called levenshtein distance
#it can be done via recursion too
#the recursion version is much more elegant yet less efficient
# https://github.com/je-suis-tm/recursion/blob/master/edit%20distance%20recursion.py

#edit distance is to minimize the steps transforming one string to another
#the way to solve this problem is very similar is to knapsack
#assume we have two strings a and b
#we build a matrix len(a)*len(b)
#however,given lists start at index zero
#our matrix should be (len(a)+1)*(len(b)+1)

#there are three different ways to transform a string
#insert,delete and replace
#we can use any of them or combined
#lets take a look at the best case first
#assume string a is string b
#we dont need to do anything
#so 0 steps would be the answer
#for the worst case
#when string a has nothing in common with string b
#we would have to replace the whole string a
#the steps become the maximum step which is max(len(a),len(b))
#for general case,the number of steps would fall between the worst and the best case
#assume we are at mth letter of string a and nth letter string b
#if we wanna get the optimal steps of transforming string a to string b
#we have to make sure at each letter transformation
#a[:m] and b[:n] have reached their optimal status
#otherwise,we could always find another combination of insert,delete and replace
#to get a "real" optimal a[:m] and b[:n]
#it would make our string transformation not so optimal any more
#it is the same logic as the optimization of knapsack problem
#after we set our logic straight
#we would take a look at three different approaches
#lets take a look at insertion
#basically we need to insert nth letter from string b into string a at nth position
#the cumulated steps we have taken should be matrix[m][n-1]+1
#matrix[m][n-1] is the steps for a[:m] to b[:n]
#for delete,it is vice versa
#the cumulated steps we have taken should be matrix[m-1][n]+1
#for replacement,it is a lil bit tricky
#there are two scenarios
#if a[m]==b[n]
#it should be matrix[m-1][n-1]
#we dont need any replacement at all
#else,it should be matrix[m-1][n-1]+1
#we replace mth letter of string a with nth letter of string b
#after we managed to understand three different approaches
#we want to take the minimum number of steps among these three approaches
#throughout the iteration of different positions of both strings
#in the end,we would get the optimal steps to transform one string to another,YAY
def edit_distance(a,b):
    
    len_a=len(a)
    len_b=len(b)
    
    #this part is to create a matrix of len(a)*len(b)
    #as lists start at index 0
    #we get a matrix of (len(a)+1)*(len(b)+1) instead
    c=[[0]*(len_b+1) for i in range(len_a+1)]
    
    for j in range(len_a+1):
        c[j][0]=j
    for k in range(len_b+1):
        c[0][k]=k
    
    #we take iterations on both string a and b
    #next,we check if a[m]==b[n]
    #if yes,no replacement needed
    #if no,replacement needed
    #we take a minimum function to see which combination would give the minimum steps
    #eventually we got what we are after
    for l in range(1,len_a+1):
        for m in range(1,len_b+1):            
            if a[l-1]==b[m-1]:
                c[l][m]=min(c[l-1][m]+1,c[l][m-1]+1,c[l-1][m-1])
            else:
                c[l][m]=min(c[l-1][m]+1,c[l][m-1]+1,c[l-1][m-1]+1)
                
    return c[len_a][len_b]

print(edit_distance('baiseé','bas'))
